File: src/main/java/me/thesourcecode/sourcebot/commands/misc/NPMCommand.java
package me.thesourcecode.sourcebot.commands.misc;

import me.thesourcecode.sourcebot.api.Source;
import me.thesourcecode.sourcebot.api.command.Command;
import me.thesourcecode.sourcebot.api.command.CommandInfo;
import me.thesourcecode.sourcebot.api.entity.SourceChannel;
import me.thesourcecode.sourcebot.api.message.alerts.ColoredAlert;
import me.thesourcecode.sourcebot.api.message.alerts.CriticalAlert;
import me.thesourcecode.sourcebot.api.utility.SourceColor;
import net.dv8tion.jda.api.MessageBuilder;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.User;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Set;

public class NPMCommand extends Command {

    private static CommandInfo INFO = new CommandInfo(
            "npm",
            "Allows you to search npmjs.org",
            "<package> (version)",
            CommandInfo.Category.GENERAL)
            .withUsageChannels(SourceChannel.COMMANDS, SourceChannel.JAVASCRIPT, SourceChannel.DISCORD, SourceChannel.DISCORD2, SourceChannel.OTHER_HELP);

    @Override
    public CommandInfo getInfo() {
        return INFO;
    }

    @Override
    public Message execute(Source source, Message message, String[] args) {
        User user = message.getAuthor();

        String version = "latest";
        String packageName = args[0].toLowerCase();

        if (args.length == 2) version = args[1].toLowerCase();

        JSONObject packageData = getJson(packageName, version);
        if (packageData == null) {
            packageName = packageName.replaceAll("`", "");
            version = version.replaceAll("`", "");

            CriticalAlert alert = new CriticalAlert();
            alert.setTitle("Uh Oh!")
                    .setDescription("I couldn't find an npm package with the name: `" + packageName + "` " +
                            (args.length == 2 ? "with the version: `" + version + "`\n" : "\n") +
                            "You can find npmjs at [npmjs.com](https://www.npmjs.com)");
            return new MessageBuilder(alert.build(user)).build();
        }

        String packageVersion = "";
        packageName = packageData.getString("name");


        String description;
        JSONArray joCollaborators;

        JSONObject time = packageData.getJSONObject("time");
        if (time.has("unpublished")) {
            description = "N/A - Package was unpublished";

            JSONObject unpublished = time.getJSONObject("unpublished");
            joCollaborators = unpublished.getJSONArray("maintainers");
            packageVersion = unpublished.getJSONObject("tags").getString("latest");
        } else {
            description = packageData.getString("description");
            joCollaborators = packageData.getJSONArray("maintainers");

            if (packageData.has("version")) {
                packageVersion = packageData.getString("version");
            } else if (packageData.has("dist-tags")) {
                packageVersion = packageData.getJSONObject("dist-tags").getString("latest");
            }
        }

        String packageLink = "https://www.npmjs.com/package/" + packageName;
        packageLink = version.equalsIgnoreCase("latest") ? packageLink : packageLink + "/v/" + packageVersion;


        StringBuilder sb = new StringBuilder();
        for (Object object : joCollaborators.toList()) {
            HashMap<String, String> collaborator = (HashMap<String, String>) object;
            sb.append("`").append(collaborator.get("name")).append("`, ");
        }
        String collaborators = sb.toString().trim();
        collaborators = collaborators.substring(0, collaborators.length() - 1);


        String packageHyperlink = time.has("unpublished") ? packageName : "[" + packageName + "](" + packageLink + ")";
        String format = "**Package Name/URL:** %s\n" +
                "**Description:** %s\n" +
                "**Version:** %s\n" +
                "**Collaborators:** %s";

        ColoredAlert alert = new ColoredAlert(SourceColor.BLUE);
        alert.setDescription(String.format(format, packageHyperlink, description, packageVersion, collaborators));
        message.getChannel().sendMessage(alert.build(user)).queue();
        return null;
    }

    private JSONObject getJson(String packageName, String version) {
        try {
            String urlString = "https://registry.npmjs.org/" + packageName;

            URL url = new URL(urlString);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();


            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            String inputLine;
            StringBuffer response = new StringBuffer();

            while ((inputLine = in.readLine()) != null) {
                response.append(inputLine);
            }
            in.close();
            connection.disconnect();

            if (!version.equalsIgnoreCase("latest")) {
                JSONObject versions = new JSONObject(response.toString()).getJSONObject("versions");
                Set<String> versionLists = versions.keySet();

                if (!versionLists.contains(version)) {
                    return null;
                }

                return versions.getJSONObject(version);
            }
            return new JSONObject(response.toString());
        } catch (Exception ex) {
            return null;
        }
    }
}

File: Extensions/Documentation/src/main/kotlin/net/sourcebot/module/documentation/commands/NPMCommand.kt
package net.sourcebot.module.documentation.commands

import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import me.hwiggy.kommander.arguments.Adapter
import me.hwiggy.kommander.arguments.Arguments
import me.hwiggy.kommander.arguments.Synopsis
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.utils.MarkdownUtil
import net.sourcebot.api.configuration.JsonSerial
import net.sourcebot.api.ifPresentOrElse
import net.sourcebot.api.response.Response
import net.sourcebot.api.response.StandardErrorResponse
import net.sourcebot.api.response.StandardInfoResponse
import net.sourcebot.module.documentation.commands.bootstrap.DocumentationCommand

class NPMCommand : DocumentationCommand(
    "npm", "Allows you to query npmjs.org"
) {
    private val registry = "https://registry.npmjs.org/"
    override val synopsis = Synopsis {
        reqParam("package", "The package to query for.", Adapter.single())
        optParam("version", "The version of the package.", Adapter.single(), "latest")
    }

    override fun execute(sender: Message, arguments: Arguments.Processed): Response {
        val packageName = arguments.required<String>("package", "You did not specify a package to query for!")
        val version = arguments.optional("version", "latest")
        val query = runCatching {
            JsonSerial.fromUrl<ObjectNode>("$registry/$packageName")
        }.getOrNull() ?: return StandardErrorResponse(
            description = "I couldn't locate package `$packageName` in the NPM Registry!"
        )
        val versionInfo = if (version != "latest") {
            query["versions"][version] ?: return StandardErrorResponse(
                description = "I couldn't locate package `$packageName` version `$version` in the NPM Registry!"
            )
        } else query
        return versionInfo["time"]["unpublished"].ifPresentOrElse({
            PackageInfo(
                packageName,
                "N/A - Package was unpublished.",
                it["tags"]["latest"].asText(),
                it["maintainers"] as ArrayNode,
                true
            )
        }, {
            PackageInfo(
                packageName,
                versionInfo["description"].asText(),
                versionInfo["version"]?.asText() ?: query["dist-tags"]["latest"].asText(),
                versionInfo["maintainers"] as ArrayNode
            )
        }).render()
    }

    private class PackageInfo(
        private val name: String,
        private val description: String,
        private val version: String,
        private val maintainers: ArrayNode,
        private val unpublished: Boolean = false
    ) {
        val link = "https://npmjs.com/package/$name/v/$version"
        fun render() = StandardInfoResponse(
            "NPM Registry - $name", """
                **Package Name / URL:** ${
                if (unpublished) name else MarkdownUtil.maskedLink(name, link)
            }
                **Description:** $description
                **Version:** $version
                **Maintainers:** ${
                maintainers.map { it["name"].asText() }.joinToString(", ") { "`$it`" }
            } 
            """.trimIndent()
        )
    }
}